doc/source/testing_pull_requests.rst

```diff
@@ -0,0 +1,178 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+- Enter inside the directory of the cloned repository in the above
+  step and run p4a command with proper args, eg::
+
+    .. codeblock:: bash
+
+        cd p4a-feature-fix-numpy
+        PYTHONPATH=. python3 -m pythonforandroid.toolchain apk \
+            --private=testapps/testapp_sqlite_openssl \
+            --dist-name=dist_test_app_python3_libs \
+            --package=org.kivy \
+            --name=test_app_python3_sqlite_openssl \
+            --version=0.1 \
+            --requirements=requests,peewee,sdl2,pyjnius,kivy,python3 \
+            --ndk-dir=/media/DEVEL/Android/android-ndk-r20 \
+            --sdk-dir=/media/DEVEL/Android/android-sdk-linux \
+            --android-api=27 \
+            --arch=arm64-v8a \
+            --permission=INTERNET \
+            --debug
+
+Things that you should know:
+
+
+    - The example above will build an testapp we will make use of the files of
+      the testapp named `testapp_sqlite_openssl.py` but we don't use the setup
+      file to build it so we must tell python-for-android what we want via
+      arguments
+    - be sure to at least edit the following arguments when running the above
+      command, since the default set in there it's unlikely that match your
+      installation:
+
+        - `--ndk-dir`: An absolute path to your android's NDK dir
+        - `--sdk-dir`: An absolute path to your android's SDK dir
+        - `--debug`: this one enables the debug mode of python-for-android,
+          which will show all log messages of the build. You can omit this one
+          but it's worth it to be mentioned, since this it's useful to us when
+          trying to find the source of the problem when things goes wrong
+    - The apk generated by the above command should be located at the root of
+      of the cloned repository, were you run the command to build the apk
+    - The testapps distributed with python-for-android are located at
+      `testapps` folder under the main folder project
+    - All the builds of python-for-android are located at
+      `~/.local/share/python-for-android`
+    - You should have a downloaded copy of the android's NDK and SDK
+
+Installing python-for-android using the github's branch of the pull request
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+- Enter inside the directory of the cloned repository mentioned in
+  `Common steps` and install it via pip, eg.::
+
+    .. codeblock:: bash
+
+        cd p4a-feature-fix-numpy
+        pip3 install . --upgrade --user
+
+- Now, go inside the `testapps` directory (we assume that you still are inside
+  the cloned repository)::
+
+    .. codeblock:: bash
+
+        cd testapps
+
+- Run the build of the apk via the freshly installed copy of python-for-android
+  by running a similar command than below::
+
+    .. code-block:: bash
+
+        python3 setup_testapp_python3_sqlite_openssl.py apk \
```

Comment:
@inclement, I need your help in here, I learned this second method from you and I have my IDE configured to do the tests without installing python-for-android as described in this method...so this should work without any p4a install right? Well...when I try to run the command from a terminal and without p4a installed, it fails with: `error: invalid command 'apk'` so I have the feeling that I miss something...what I'm missing?

Reply:
You do need to install python-for-android for this to work - if you look in the p4a setup.py, the `apk` argument to setup.py is one of the things that gets installed as part of p4a.

Reply:
Aaa, thanks for the clarification...with your info I just found why I can do this without install p4a from my ide

Seeing my forked python-for-android files, I have an directory that we don't have committed `python_for_android.egg-info` (because at some point I installed p4a from my fork and the `egg folder` was generated... I never took care to delete that folder)...that folder seems that is the key, because it contains the `entrypoints.txt` ...the mentioned `apk` is in there...If I remove this file I got the exact error I mentioned before...

So i think that we just found another test method without install...I will try to write this one...

thanks @inclement :smile:

Reply:
Ok, I've been investigating a little this thing, and I can say that I realized that, some time ago, I hacked my `site-packages` for the python version that I usually use to test p4a  (**I didn't remember that**)...so it's really necessary to install the package. So better that we stick  to the words I wrote for this method, but I will explain what the hell I did to make it work, which I don't think that we should put in the documentation, but still maybe someone want to do this:

  - I installed the python-for-android package from my development fork of p4a
  - I removed the `pythonforandroid` folder from the installed package in site-packages
  - I created a symlink who points to my `pythonforandroid` folder of my fork of p4a

Doing it this way I can use this method without doing any install (**of course...it's always installed**)...I just switch to the branch I want to test and execute the command for the testapp. So this method it's interesting for development (the best way to apply this is to have an python's virtualenv for only p4a, so you don't mess up with the system's python)

Reply:
Oh yes I've used the symlink approach quite a lot also. Pretty useful, but indeed better leaving this hack undocumented :smile: